services/121-service/src/transaction-jobs/services/transaction-jobs-onafriq.service.ts

```diff
@@ -0,0 +1,122 @@
+import { OnafriqService } from '@121-service/src/payments/fsp-integration/onafriq/onafriq.service';
+import { TransactionStatusEnum } from '@121-service/src/payments/transactions/enums/transaction-status.enum';
+import { TransactionScopedRepository } from '@121-service/src/payments/transactions/transaction.repository';
+import { ScopedRepository } from '@121-service/src/scoped.repository';
+import { TransactionJobsHelperService } from '@121-service/src/transaction-jobs/services/transaction-jobs-helper.service';
+import { OnafriqTransactionJobDto } from '@121-service/src/transaction-queues/dto/onafriq-transaction-job.dto';
+import { getScopedRepositoryProviderName } from '@121-service/src/utils/scope/createScopedRepositoryProvider.helper';
+import { generateUUIDFromSeed } from '@121-service/src/utils/uuid.helpers';
+
+@Injectable()
+export class TransactionJobsOnafriqService {
+  constructor(
+    private readonly onafriqService: OnafriqService,
+    @Inject(getScopedRepositoryProviderName(OnafriqTransactionEntity))
+    private readonly onafriqTransactionScopedRepository: ScopedRepository<OnafriqTransactionEntity>,
+    private readonly transactionScopedRepository: TransactionScopedRepository,
+    private readonly transactionJobsHelperService: TransactionJobsHelperService,
+  ) {}
+
+  public async processOnafriqTransactionJob(
+    transactionJob: OnafriqTransactionJobDto,
+  ): Promise<void> {
+    // 1. Create idempotency key
+    const registration =
+      await this.transactionJobsHelperService.getRegistrationOrThrow(
+        transactionJob.referenceId,
+      );
+    const failedTransactionsCount =
+      await this.transactionScopedRepository.count({
+        where: {
+          registrationId: Equal(registration.id),
+          payment: Equal(transactionJob.paymentNumber),
+          status: Equal(TransactionStatusEnum.error),
+        },
+      });
+    // thirdPartyTransId is generated using: (referenceId + paymentNr + failedTransactionsCount)
+    // Using this count to generate the thirdPartyTransId ensures that on:
+    // a. Payment retry, a new thirdPartyTransId is generated, which will not be blocked by Onafriq API, as desired.
+    // b. Queue retry: on queue retry, the same thirdPartyTransId is generated, which will be blocked by Onafriq API, as desired.
+    const thirdPartyTransId = generateUUIDFromSeed(
+      `ReferenceId=${transactionJob.referenceId},PaymentNumber=${transactionJob.paymentNumber},Attempt=${failedTransactionsCount}`,
+    );
+
+    // 2. Check for existing Onafriq Transaction with the same thirdPartyTransId, because that means this job has already been (partly) processed. In case of a server crash, jobs that were in process are processed again.
+    let onafriqTransaction =
+      await this.onafriqTransactionScopedRepository.findOne({
+        where: {
+          thirdPartyTransId: Equal(thirdPartyTransId),
+        },
+      });
```

Comment:
Maybe it's nice to to unit this flow if there already is an onafriqTransaction? As that is not covered by integration tests now. And you can mock it well I think for a unit test

Reply:
good idea. added.